pychiquito/fibonacci.py

```diff
@@ -36,99 +24,45 @@ def __init__(self: Fibonacci):
-            ctx.assign(circuit.b, F(b_value))
-            ctx.assign(self.c, F(a_value + b_value))
+    def wg(self: FiboStep, args: Tuple[int, int]):
+        a_value, b_value = args
+        self.assign(self.circuit.a, F(a_value))
+        self.assign(self.circuit.b, F(b_value))
+        self.assign(self.c, F(a_value + b_value))
 
-        super().wg(wg_def)
 
-
-class FiboLastStep(StepTypeContext):
+class FiboLastStep(StepType):
     def setup(self: FiboLastStep):
         self.c = self.internal("c")
         self.constr(eq(self.circuit.a + self.circuit.b, self.c))
 
-    def wg(self: FiboLastStep, circuit: Fibonacci):
-        # Any instead of Args
-        def wg_def(ctx: StepInstance, values: Tuple[int, int]):
-            a_value, b_value = values
-            print(
-                f"fib last step wg: {a_value}, {b_value}, {a_value + b_value}\n")
-            ctx.assign(circuit.a, F(a_value))
-            ctx.assign(circuit.b, F(b_value))
-            ctx.assign(self.c, F(a_value + b_value))
-
-        super().wg(wg_def)
+    def wg(self: FiboLastStep, values=Tuple[int, int]):
+        a_value, b_value = values
+        self.assign(self.circuit.a, F(a_value))
+        self.assign(self.circuit.b, F(b_value))
+        self.assign(self.c, F(a_value + b_value))
 
 
-# Print Circuit
 fibo = Fibonacci()
-fibo.trace()
-print("Print Circuit using custom __str__ method in python:")
-print(fibo.circuit)
-print("Print Circuit using __json__ method in python:")
-
-
-class CustomEncoder(json.JSONEncoder):
-    def default(self, obj):
-        if hasattr(obj, "__json__"):
-            return obj.__json__()
-        return super().default(obj)
-
-
-# Print Circuit
-print("Print Circuit using custom __str__ method in python:")
-print(fibo.circuit)
-print("Print Circuit using __json__ method in python:")
-circuit_json = json.dumps(fibo.circuit, cls=CustomEncoder, indent=4)
-print(circuit_json)
-
-# Print TraceWitness
-trace_generator = TraceGenerator(fibo.circuit.trace)
-trace_witness = trace_generator.generate(None)
-print("Print TraceWitness using custom __str__ method in python:")
-print(trace_witness)
-print("Print TraceWitness using __json__ method in python:")
-trace_witness_json = json.dumps(trace_witness, cls=CustomEncoder, indent=4)
-print(trace_witness_json)
-
-# Rust bindings for Circuit
-print("Call rust bindings, parse json to Chiquito ast, and print using Debug trait:")
-rust_chiquito.convert_and_print_ast(circuit_json)
-print(
-    "Call rust bindings, parse json to Chiquito TraceWitness, and print using Debug trait:"
-)
-rust_chiquito.convert_and_print_trace_witness(trace_witness_json)
-print("Parse json to Chiquito Halo2, and obtain UUID:")
-ast_uuid: int = rust_chiquito.ast_to_halo2(circuit_json)
-print("Verify ciruit with ast uuid and trace witness json:")
-rust_chiquito.verify_proof(trace_witness_json, ast_uuid)
+fibo_witness = fibo.gen_witness(None)
+fibo.ast_to_halo2()
+fibo.verify_proof(fibo_witness)
```

Comment:
perhaps you should assert what should be the result of this call should be

Reply:
You can print in the example, what I was saying is to not have the printing in the library

Reply:
Yes I agree. Again the UUID is preventing an assertion style test, but let me know if you have any better ideas. I think I can technically loop through all elements and replace UUID style elements before doing assertion, but looping through a complex structure isn't intuitive and this can make the code quite long. Leaving it as is for now as `verify_proof` technically tests both the AST and the witness.